Source/Scene/ModelExperimental/MaterialPipelineStage.js

```diff
@@ -0,0 +1,513 @@
+import AlphaMode from "../AlphaMode.js";
+import LightingModel from "./LightingModel.js";
+import MaterialStageFS from "../../Shaders/ModelExperimental/MaterialStageFS.js";
+
+/**
+ * The material pipeline stage processes textures and other uniforms needed
+ * to render a primitive. This handles the following material types:
+ * <ul>
+ *   <li>Basic glTF materials (PBR metallic roughness model)</li>
+ *   <li>The `KHR_materials_pbrSpecularGlossiness` glTF extension</li>
+ *   <li>The `KHR_materials_unlit` glTF extension</li>
+ * </ul>
+ *
+ * @private
+ */
+export default function MaterialPipelineStage() {}
+
+/**
+ * Process a primitive. This modifies the following parts of the render
+ * resources:
+ * <ul>
+ *   <li>Modifies the shader to include the material processing stage</li>
+ *   <li>Modifies the shader to include additional uniforms for textures and other rendering details</li>
+ *   <li>Modifies the lighting options to set either PBR or unlit lighting</li>
+ *   <li>Sets the render state for back-face culling</li>
+ * </ul>
+ * @param {PrimitiveRenderResources} renderResources
+ * @param {ModelComponents.Primitive} primitive The primitive to be rendered
+ * @private
+ */
+MaterialPipelineStage.process = function (renderResources, primitive) {
+  var material = primitive.material;
+
+  var uniformMap = renderResources.uniformMap;
+  var shaderBuilder = renderResources.shaderBuilder;
+
+  processMaterialUniforms(material, uniformMap, shaderBuilder);
+
+  if (defined(material.specularGlossiness)) {
+    processSpecularGlossinessUniforms(material, uniformMap, shaderBuilder);
+  } else {
+    processMetallicRoughnessUniforms(material, uniformMap, shaderBuilder);
+  }
+
+  var lightingOptions = renderResources.lightingOptions;
+  if (material.unlit) {
+    lightingOptions.lightingModel = LightingModel.UNLIT;
+  } else {
+    lightingOptions.lightingModel = LightingModel.PBR;
+  }
+
+  // Configure back-face culling
+  renderResources.renderStateOptions.cull = {
+    enabled: !material.doubleSided,
+  };
+
+  addAlphaUniforms(material, uniformMap, shaderBuilder);
+
+  shaderBuilder.addFragmentLines([MaterialStageFS]);
+};
+
+function processMaterialUniforms(material, uniformMap, shaderBuilder) {
+  var texCoordIndex;
+
+  var emissiveTexture = material.emissiveTexture;
+  if (defined(emissiveTexture)) {
+    shaderBuilder.addUniform(
+      "sampler2D",
+      "u_emissiveTexture",
+      ShaderDestination.FRAGMENT
+    );
+    uniformMap.u_emissiveTexture = function () {
+      return material.emissiveTexture.texture;
+    };
+    texCoordIndex = emissiveTexture.texCoord;
+    shaderBuilder.addDefine(
+      "HAS_EMISSIVE_TEXTURE",
+      undefined,
+      ShaderDestination.FRAGMENT
+    );
+    shaderBuilder.addDefine(
+      "TEXCOORD_EMISSIVE",
+      "v_texCoord_" + texCoordIndex,
+      ShaderDestination.FRAGMENT
+    );
+
+    if (defined(emissiveTexture.transform)) {
+      shaderBuilder.addDefine(
```

Comment:
I need to find (or make) a model with texture transforms for emissive, normal, etc.

Reply:
Nevermind, when I refactored the texture processing code, I added some unit tests that cover the texture transform processing.